Orbitersdk/samples/ProjectApollo/src_sys/thread.cpp

```diff
@@ -23,22 +23,37 @@
   **************************************************************************/
 
 #include "thread.h"
+#ifdef _WIN32
+#include <windows.h>
+#endif
 
-void Runnable::Kill ()
-{
-    thread.WaitForDeath ();
+KillableWorker::~KillableWorker() {
+    Kill();
 }
 
-Runnable::Runnable ():
-#pragma warning(disable: 4355) // 'this' used before initialized
-  thread (ThreadEntry, this)
-#pragma warning(default: 4355)
-{
+void KillableWorker::Start(std::function<void()>f) {
+    Kill();
+    m_running = true;
+    m_thread = std::thread([this, f] {
+                    f();
+                    m_running = false;
+               });
 }
 
-DWORD WINAPI Runnable::ThreadEntry (void* arg)
-{
-    Runnable * pRunnable = (Runnable *) arg;
-    pRunnable->Run();
-    return 0;
+// Some RTCC algorithms can loop forever if there is a convergence problem,
+// so we need a way to kill the worker thread.
+// Unfortunately there's no standard way to do that so we have to revert to
+// native thread API to do it
+void KillableWorker::Kill() {
```

Comment:
This method leaks resources. I would add a comment that notifies the user of that fact so that this doesn't get used without explicitly knowing you are causing a memory leak. This should only be used if there is no other way to safely terminate a thread.

I do agree that there is no platform independent approach that covers this nicely.

Reply:
Comment added, hope it's not too dramatic :grin:

Reply:
I think that gets the message across. Thanks!